File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/OptionsTestsReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if GAS || INTEL || MASM || NASM || FAST_FMT
using System;
using System.Collections.Generic;
using System.IO;
using Iced.Intel;

namespace Iced.UnitTests.Intel.FormatterTests {
	static class OptionsTestsReader {
		public static IEnumerable<OptionsTestCase> ReadFile(string filename, HashSet<int> ignored) {
			int lineNo = 0;
			int testCaseNo = 0;
			foreach (var line in File.ReadLines(filename)) {
				lineNo++;
				if (line.Length == 0 || line[0] == '#')
					continue;
				OptionsTestCase? testCase;
				try {
					testCase = ReadTestCase(line, lineNo);
				}
				catch (Exception ex) {
					throw new InvalidOperationException($"Error parsing options test case file '{filename}', line {lineNo}: {ex.Message}");
				}
				if (testCase.HasValue)
					yield return testCase.GetValueOrDefault();
				else
					ignored.Add(testCaseNo);
				testCaseNo++;
			}
		}

		static readonly char[] seps = new char[] { ',' };
		static readonly char[] optsseps = new char[] { ' ' };
		static OptionsTestCase? ReadTestCase(string line, int lineNo) {
			var parts = line.Split(seps);
			if (parts.Length != 4)
				throw new InvalidOperationException($"Invalid number of commas ({parts.Length - 1} commas)");

			int bitness = NumberConverter.ToInt32(parts[0].Trim());
			var ip = bitness switch {
				16 => DecoderConstants.DEFAULT_IP16,
				32 => DecoderConstants.DEFAULT_IP32,
				64 => DecoderConstants.DEFAULT_IP64,
				_ => throw new InvalidOperationException(),
			};
			var hexBytes = parts[1].Trim();
			HexUtils.ToByteArray(hexBytes);
			var codeStr = parts[2].Trim();
			if (CodeUtils.IsIgnored(codeStr))
				return null;
			var code = ToCode(codeStr);

			var properties = new List<(OptionsProps property, object value)>();
			foreach (var part in parts[3].Split(optsseps, StringSplitOptions.RemoveEmptyEntries))
				properties.Add(OptionsParser.ParseOption(part));

			return new OptionsTestCase(bitness, hexBytes, ip, code, properties);
		}

		static Code ToCode(string value) {
			if (!ToEnumConverter.TryCode(value, out var code))
				throw new InvalidOperationException($"Invalid Code value: '{value}'");
			return code;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/OptionsTestsReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import java.util.ArrayList;
import java.util.HashSet;

import com.github.icedland.iced.x86.CodeUtils;
import com.github.icedland.iced.x86.DecoderConstants;
import com.github.icedland.iced.x86.FileUtils;
import com.github.icedland.iced.x86.HexUtils;
import com.github.icedland.iced.x86.NumberConverter;
import com.github.icedland.iced.x86.StringUtils2;
import com.github.icedland.iced.x86.ToCode;

final class OptionsTestsReader {
	public static ArrayList<OptionsTestCase> readFile(String filename, HashSet<Integer> ignored) {
		ArrayList<OptionsTestCase> result = new ArrayList<OptionsTestCase>();
		ArrayList<String> lines = FileUtils.readAllLines(filename);
		int lineNo = 0;
		int testCaseNo = 0;
		for (String line : lines) {
			lineNo++;
			if (line.length() == 0 || line.charAt(0) == '#')
				continue;
			OptionsTestCase testCase;
			try {
				testCase = readTestCase(line, lineNo);
			} catch (Exception ex) {
				throw new UnsupportedOperationException(
						String.format("Error parsing options test case file '%s', line %d: %s", filename, lineNo, ex.getMessage()));
			}
			if (testCase != null)
				result.add(testCase);
			else
				ignored.add(testCaseNo);
			testCaseNo++;
		}
		return result;
	}

	static OptionsTestCase readTestCase(String line, int lineNo) {
		String[] parts = StringUtils2.split(line, ",");
		if (parts.length != 4)
			throw new UnsupportedOperationException(String.format("Invalid number of commas (%d commas)", parts.length - 1));

		int bitness = NumberConverter.toInt32(parts[0].trim());
		long ip;
		switch (bitness) {
		case 16:
			ip = DecoderConstants.DEFAULT_IP16;
			break;
		case 32:
			ip = DecoderConstants.DEFAULT_IP32;
			break;
		case 64:
			ip = DecoderConstants.DEFAULT_IP64;
			break;
		default:
			throw new UnsupportedOperationException();
		}
		String hexBytes = parts[1].trim();
		HexUtils.toByteArray(hexBytes);
		String codeStr = parts[2].trim();
		if (CodeUtils.isIgnored(codeStr))
			return null;
		int code = ToCode.get(codeStr);

		ArrayList<OptionsTestCase.Info> properties = new ArrayList<OptionsTestCase.Info>();
		for (String part : parts[3].split(" ")) {
			if (part.equals(""))
				continue;
			properties.add(OptionsParser.parseOption(part));
		}

		return new OptionsTestCase(bitness, hexBytes, ip, code, properties.toArray(new OptionsTestCase.Info[0]));
	}
}
